### PR TITLE
Add a new way to hold structs in `PolymorphicValue`

### PR DESCRIPTION
**Pull request by zasdfgbnm**
See note `[Struct Support in PolymorphicValue]` for description, and the new test `PolymorphicValueTest.Struct` for examples.

### CODE REVIEW

File: csrc/struct.inl
Comment (zasdfgbnm) [inline csrc/struct.inl:1]: Note to myself: add `.inl` check to linters

File: csrc/polymorphic_value.h [hunk 1]
```diff
@@ -236,8 +236,50 @@ inline std::ostream& operator<<(std::ostream& os, const Pointer& ptr) {
   return os;
 }
 
+struct Struct;
+struct Accessor;
+struct StructType;
+
+class StructHolder {
```
Comment (zasdfgbnm) [inline csrc/polymorphic_value.h:243]:
```suggestion
class StructHandle {
```

File: csrc/struct.inl [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <string>
+
+namespace nvfuser {
```
Comment (zasdfgbnm) [inline csrc/struct.inl:14]: Add note describing the big picture of nvFuser struct support.

File: csrc/ir/builder.cpp [hunk 3]
```diff
@@ -258,21 +258,6 @@ Val* IrBuilder::metadataExpr(TensorView* tv) {
   return tv->fusion()->metadataOf(tv);
 }
 
-Val* IrBuilder::structExpr(
```
Comment (zasdfgbnm) [inline csrc/ir/builder.cpp:261]: Moved to header file, because it now needs a template parameter

**Comment by zasdfgbnm** [comment]
!build

**Review by naoyam — APPROVED** [review]
LGTM

File: csrc/struct.inl
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/struct.inl:1]: `inl` seems to be the standard suffix for the case where you feel a header is too long, and want to put part of it in a separate file:
https://www.oreilly.com/library/view/c-cookbook/0596007612/ch02s06.html

[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/struct.inl:1]: The example in the linked page makes sense. It separates declarations and implementations. That doesn't seem to be the case with `Struct`?

[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/struct.inl:1]: For the case of `Struct`, I am considering `StructHandle` as an interface (with `PolymorphicValue`) for struct implementation, but `Struct` and `Accessor` and `NotImplementedStruct` as implementation detail for struct. I agree that this separation of interface vs detail is not accurate, but still, I think:
1. It's better to put them in a separate file.
2. I can not think of a better name than `.inl` for that separate file.

I can name it to `struct.h`, but I have never seen one header includes another header in its end for pulling in more thing. But I have seen the name `inl` being used a few times. There is an example for using `*-inl.h` for implementations like inline function and `numeric_limits`.

[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/struct.inl:1]: That makes sense. Thanks for the explanation.

**Pull request merged**